### PR TITLE
UIA Operation Abstraction lib: implement support for CallExtension / IsExtensionSupported

### PR DESCRIPTION
**Pull request by michaelDCurran**
UI Automation recently introduced the concept of custom patterns. This is the ability for a provider to expose custom methods, identified by a GUID, that a client an execute via UIA remote Operations. For example, MS Word might expose a custom pattern for moving by sentence - something that UIA by itself does not support.

This PR implements the ability for a client to execute these custom methods (extensions) by using the UIA Operation Abstraction library. Support for custom patterns has already been introduced to the Microsoft::UI::UIAutomation winrt library.

Specific changes:
* Added a new UiaGuid class, which represents either a local or remote GUID.
	* It can be constructed locally from a winrt::guid.
	* It can be assigned to, and compaired with == / !=.
	* It exposes LookupAnnotationType and LookupPropertyId methods, which work both remotely and locally.
* Implement UiaElement::IsExtensionSupported: this method takes a UiaGuid as an argument, and returns a UiaBool denoting if the extension identified by the given GUID is supported by the remote provider.
	* If remote, this method calls AutomationRemoteElement::IsExtensionSupported.
	* This method has no local equivilant and therefore throws NotImplemented if executed locally.
* Implemented UiaElement::CallExtension: this method takes a UiaGuid, plus a variable number of further arguments of any Uia* type, and executes the extension identified by the given GUID, passing the given arguments as its in/out arguments.
	* If remote, this method calls AutomationRemoteElement::CallExtension.
	* This method has no local equivilant and therefore throws NotImplemented if executed locally. 

No tests have been written yet. These will be hard to write unless they can be run against a specific app that exposes 1 or more custom patterns. 
Manually I have tested UiaGuid::LookupAnnotationType / UiaGuid::LookupPropertyId. But I am so far unable to test UiaElement::IsExtensionSupported / UIAElement::CallExtension until I have an app that actually implements 1 or more custom patterns...

### CODE REVIEW

**Comment by michaelDCurran** [comment]
I have been able to manually confirm that these changes work, at least for my use cases, by testing them against Microsoft Word's custom patterns implementation, from within the NVDA screen reader. However, I really am not sure how this can be tested specifically within this project as access to Microsoft Word, or something that implemented custom patterns, would be required on the build system.

**Comment by beweedon** [comment]
Sorry for us ignoring this PR for so long! I'll give it a look next week. @mavangur @zhayuli @pengyuanjin for FYI.

Honestly, we haven't even had the time to get our tests running in CI yet. There's a general issue with launching apps and we haven't really gotten to look at why. I'll give some thought as to what we can do with regard to automated testing here.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h
Comment (pengyuanjin) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:2185]: Do we need these comment lines?

**Comment by michaelDCurran** [comment]
I tried to add support for GUIds to UIAVariant, but did not succeed.  I 
am unsure as to what vt type should be used, or if it it is at all 
possible to hold a GUID in a VARIANT structure at all these days?


I'm happy to either remove these commented lines, or if anyone has 
further info on how to do this properly, please share and I'll try 
adding that support again.

Message ID: 
***@***.***>

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp
Comment (beweedon) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:727]: **nit:** This should be indented in by 4 more spaces.

File: src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstraction.g.h
Comment (beweedon) [inline src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstraction.g.h:2171]: **nit:** Should be spelled "equivalent"

**Comment by beweedon** [comment]
@michaelDCurran I'd say as long as we don't need GUIDs in UiaVariant for this PR, we can remove those comment lines. It might make sense to store GUIDs in VT_SAFEARRAY in the variant, though, when we want to implement that.

Also, we could add support for extension tests by creating a provider that supports custom extensions, and loading it with the `ModernApp` test utility (or creating a Win32 executable). If you feel you're up to that in this PR feel free (and of course ask us if you need help!), but if not I've created #86 to track adding those tests in the future. I can try to get around to confirming your manual testing next week.

**Comment by beweedon** [comment]
I'm really sorry for all the slow responses here!

My remaining comment is can we add tests for `UiaGuid`? We don't currently have a huge amount of tests for each individual type, but I figure we should aim for good test coverage on new types. So basically calling all its methods, operating locally and remotely, etc.

Otherwise, this looks great!!!

**Comment by beweedon** [comment]
@michaelDCurran just in case you missed these notifications.

**Comment by michaelDCurran** [comment]
Sorry for delay on this. Hopefully I should find time in next week or so 
to write UiaGuid tests.

**Comment by beweedon** [comment]
No need to apologize I just wanted to make sure you didn't think you were waiting on me :)

**Comment by michaelDCurran** [comment]
@beweedon  I'm afraid that I can't think of any ways I can write remote tests for UiaGuid.
* There is no method (other than UIA custom extensions) that take a remote guid
* There are no UIA methods that give back a guid
* A guid to / from a custom property or annotation only makes sense if there are custom annotations or properties already registered in the remote provider.

In other words, to create any meaningful tests for UiaGuid, (just like for custom extensions) we'd need to write a custom provider or test with an MS Office product such as Microsoft Word.
I don't believe that I am qualified at the moment to pull that one off at this stage. Plus, some of the other additions we have talked about are probably more high priority for me. Thankfully those ones are actually quite easy to write tests for ;)
I'll leave it up to you then as to if you wish to merge this with no custom provider / tests. I'm already using this code real-world in NVDA 2022.1 via my fork michaeldcurran/microsoft-ui-uiautomation with the code from this pr merged.

**Comment by beweedon** [comment]
@michaelDCurran I was thinking mostly about testing the simple stuff just to make sure it works.

- Test that `UiaGuid` can go from local to remote back to local again.
- Test that `UiaGuid` assignment and equality work.
- Test that `LookupPropertyId` works if you set the `UiaGuid` to one of the known property GUIDs in UIAutomationCoreApi.h (e.g. `Name_Property_GUID` should correspond to `UIA_NamePropertyId`).

I'm glad you're not blocked, though :). If you don't want to add the tests, one of us can look into that at some point and get this merged in.

**Comment by michaelDCurran** [comment]
@beweedon  I've now added tests for UiaGuid, including equality checks, and looking up property IDs and annotation type IDs.

**Comment by beweedon** [comment]
@michaelDCurran This looks great to me! I'm going to see if other reviews filter in before merging.

**Review by beweedon — APPROVED** [review]

**Review by csauls — APPROVED** [review]

**Pull request merged**
